presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -585,17 +596,136 @@ protected Node visitExpression(Expression node, Void context)
+                return new FunctionCall(CARDINALITY, ImmutableList.of(mergedHll));
+            }
+
+            boolean canRewrite(Expression expression)
+            {
+                if (!(expression instanceof FunctionCall)) {
+                    return false;
+                }
+
+                FunctionCall functionCall = (FunctionCall) expression;
+                if (functionCall.isDistinct() || functionCall.getWindow().isPresent() || functionCall.getOrderBy().isPresent()
+                        || functionCall.isIgnoreNulls() || functionCall.getFilter().isPresent()) {
+                    return false;
+                }
+
+                QualifiedName functionName = functionCall.getName();
+                List<Expression> expressions = functionCall.getArguments();
+
+                if (!NONSTANDARD_REWRITE_FUNCTIONS.contains(functionName) || expressions.size() > 1 || !(expressions.get(0) instanceof Identifier)) {
+                    return false;
+                }
+                Identifier baseTableColumn = (Identifier) expressions.get(0);
+
+                return validateFunctions.get(functionName).apply(baseTableColumn);
+            }
+
+            private boolean validateApproxDistinct(Identifier baseTableColumn)
+            {
+                Cast cast = new Cast(new FunctionCall(APPROX_SET, ImmutableList.of(baseTableColumn)), VARBINARY);
+                return materializedViewInfo.getBaseToViewColumnMap().containsKey(cast);
+            }
+
+            private boolean validateAvg(Identifier baseTableColumn)
+            {
+                // To create average from MV, we need  either sum and avg, or sum and count
+                // TODO: add support for count and avg
+                boolean containsSum = getDerivedFromFunctionCall(SUM, baseTableColumn).isPresent();
+                boolean containsAvg = getDerivedFromFunctionCall(AVG, baseTableColumn).isPresent();
+                boolean containsCount = getDerivedFromFunctionCall(COUNT, baseTableColumn).isPresent();
+
+                return (containsSum && containsAvg) || (containsSum && containsCount);
+            }
+
+            /**
+             * Gets the column from the materialized view derived from provided function name and argument
+             */
+            private Optional<Identifier> getDerivedFromFunctionCall(QualifiedName functionName, Identifier baseTableColumn)
+            {
+                return Optional.ofNullable(materializedViewInfo.getBaseToViewColumnMap().get(new FunctionCall(functionName, ImmutableList.of(baseTableColumn))));
+            }
+        }
+
         @Override
         protected Node visitFunctionCall(FunctionCall node, Void context)
         {
             ImmutableList.Builder<Expression> rewrittenArguments = ImmutableList.builder();
 
+            Optional<Expression> functionCall = new NonAssociativeFunctionHandler().rewriteFunction(node);
```

Comment:
Mark rewriteFunction as static, so we don't have to do new NonAssociativeFunctionHandler() multiple times